internal/errorlint/errorlint.go

```diff
@@ -0,0 +1,191 @@
+	pass := reporter.NewPass(name, _pass, reporter.Warn())
+	for _, astFile := range pass.Files {
+		// Ignore generated files and test files.
+		if common.IsGenerated(astFile) || common.IsTestFile(pass.Pass, astFile) {
+			continue
+		}
+		newFile := &file{f: astFile}
+		lintMessageStrings(newFile, pass)
+	}
+
+	return nil, nil
+}
+
+// lintMessageStrings examines error/trace/log message strings for capitalization and valid ending
+func lintMessageStrings(file *file, pass *reporter.Pass) {
+	file.walk(func(node ast.Node) bool {
+		call, ok := node.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+
+		if isErrorPackage(call.Fun) || len(call.Args) < 1 {
+			return true
+		}
+
+		msgIndex := 1
+		if isDotInPkg(call.Fun, "errors", "New") || isDotInPkg(call.Fun, "fmt", "Errorf") ||
+			isDotInPkg(call.Fun, "errors", "Errorf") {
+			msgIndex = 0
+		}
+
+		msg, ok := call.Args[msgIndex].(*ast.BasicLit)
+		if !ok || msg.Kind != token.STRING {
+			return true
+		}
+
+		msgString, err := strconv.Unquote(msg.Value)
+		if err != nil {
+			return false
+		}
+
+		if msgString == "" {
+			return true
+		}
+		errormsg := getErrorMessage(msgString)
+		if errormsg != "" {
+			pkgName := getPkgName(call.Fun)
+			pass.Reportf(node.Pos(), "%s "+errormsg, pkgName)
+		}
+		return true
+	})
+}
+
+// isErrorPackage checks if the ast.Expr package matches the error/fmt/trace/log packages for linter
+func isErrorPackage(expr ast.Expr) bool {
+	return !isDotInPkg(expr, "errors", "New") && !isDotInPkg(expr, "errors", "Wrap") &&
+		!isDotInPkg(expr, "errors", "Wrapf") && !isDotInPkg(expr, "log", "Warn") &&
+		!isDotInPkg(expr, "log", "Info") && !isDotInPkg(expr, "log", "Error") &&
+		!isDotInPkg(expr, "trace", "StartSpan") && !isDotInPkg(expr, "trace", "StartCall") &&
+		!isDotInPkg(expr, "fmt", "Errorf") && !isDotInPkg(expr, "errors", "Errorf") &&
+		!isDotInPkg(expr, "errors", "WithMessage") && !isDotInPkg(expr, "errors", "WithMessagef")
+}
+
+// getErrorMessage returns message based on whether it has capitalization, punctuation or not
+func getErrorMessage(msg string) string {
+	isCap, isPunct := isStringFormatted(msg)
+	var errormsg string
+	switch {
+	case isCap && isPunct:
+		errormsg = "message should not be capitalized and should not end with punctuation"
+	case isCap:
+		errormsg = "message should not be capitalized"
+	case isPunct:
+		errormsg = "message should not end with punctuation"
+	default:
+		errormsg = ""
+	}
+
+	return errormsg
+}
+
+// isIdent checks if ident string is equal to ast.Ident name
+func isIdent(expr ast.Expr, ident string) bool {
+	id, ok := expr.(*ast.Ident)
+	return ok && id.Name == ident
+}
+
+// hasDotInPkg checks if pkg.function format is followed
+func isDotInPkg(expr ast.Expr, pkg, name string) bool {
+	sel, ok := expr.(*ast.SelectorExpr)
+	return ok && isIdent(sel.X, pkg) && isIdent(sel.Sel, name)
+}
+
+// getPkgName returns package name errors/fmt/log/trace
+func getPkgName(expr ast.Expr) string {
+	sel, ok := expr.(*ast.SelectorExpr)
```

Comment:
Short-circuit here on `!ok`? It saves you a check below:
```suggestion
	sel, ok := expr.(*ast.SelectorExpr)
	if !ok {
		return ""	
	}
```

Reply:
added